### PR TITLE
Add more leaf fields to logsdb data generator

### PR DESCRIPTION
**Pull request by lkts**
This PR adds all numeric field types to data generator.

In scope of this work i realized that current approach of having a single `Arbitrary` interface is approaching its limits. It has also came up in a review previously. One problem is that every time a new kind of generated data is introduced (like in this PR), `Arbitrary` needs to be changed therefore all consumers needs to change (right now its only the snapshot test but it is already pretty annoying). Another problem is simply the size of implementation when everything related to random generation is in one place. 

After some experiments i landed on approach in this PR which has some nice properties:
1. Adding a new type of data is a purely additive change, what previously worked continues to work. Same with making a change to a specific type of data.
3. Flexibility - request, response, and handler can be whatever you want and do whatever you want. E.g. you can imagine a new handler for mapping parameters.
4. Continuing the above point - code structure is flexible. Similar things like primitive data type generation can be (and are) grouped together, unrelated things can be split into different handlers.
5. Overrides using override handlers are easy and precise - you override only behavior you care about and everything else stays the same. Overrides are not related to default implementations in any way and again are strictly additive changes.

It is verbose and has this maybe-controversial parallel Request-Response class hierarchy. I do think it's an improvement and i think properties above are a big deal though.

### CODE REVIEW

File: modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/logsdb/qa/StandardVersusLogsIndexModeRandomDataChallengeRestIT.java [hunk 1]
```diff
@@ -39,22 +41,22 @@ public StandardVersusLogsIndexModeRandomDataChallengeRestIT() {
...
                     // and equality check fails because arrays are sorted in synthetic source.
-                    @Override
-                    public boolean generateArrayOfValues() {
-                        return false;
+                    if (request instanceof DataSourceRequest.ArrayWrapper) {
+                        return new DataSourceResponse.ArrayWrapper(Function.identity());
                     }
-                })
+
+                    return new DataSourceResponse.NotMatched();
```
Comment (kkrik-es) [inline modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/logsdb/qa/StandardVersusLogsIndexModeRandomDataChallengeRestIT.java:58]: Can this if-sequence be replaced by inheritance?

File: test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/datasource/DataSourceRequest.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.logsdb.datageneration.datasource;
+
+import org.elasticsearch.logsdb.datageneration.DataGeneratorSpecification;
+
+public interface DataSourceRequest {
```
Comment (kkrik-es) [inline test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/datasource/DataSourceRequest.java:13]: Oh hey protocol buffers :P

File: test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/datasource/DataSourceResponse.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+    record HalfFloatGenerator(Supplier<Float> generator) implements DataSourceResponse {}
+
+    record StringGenerator(Supplier<String> generator) implements DataSourceResponse {}
+
+    record NullWrapper(Function<Supplier<Object>, Supplier<Object>> wrapper) implements DataSourceResponse {}
+
+    record ArrayWrapper(Function<Supplier<Object>, Supplier<Object>> wrapper) implements DataSourceResponse {}
+
+    interface ChildFieldGenerator extends DataSourceResponse {
```
Comment (kkrik-es) [inline test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/datasource/DataSourceResponse.java:47]: I assume we'll be adding generateArrayField, generateArrayObject here?

File: test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/datasource/DefaultObjectGenerationHandler.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                @Override
+                public FieldType generateFieldType() {
+                    return randomFrom(FieldType.values());
+                }
+            };
+        }
+
+        if (request instanceof DataSourceRequest.ObjectArrayGenerator) {
+            return new DataSourceResponse.ObjectArrayGenerator(() -> {
+                if (ESTestCase.randomBoolean()) {
```
Comment (kkrik-es) [inline test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/datasource/DefaultObjectGenerationHandler.java:56]: Use `.usually()`?

File: test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/datasource/DefaultPrimitiveTypesHandler.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+ */
+
+package org.elasticsearch.logsdb.datageneration.datasource;
+
+import org.elasticsearch.test.ESTestCase;
+
+import java.math.BigInteger;
+
+public class DefaultPrimitiveTypesHandler implements DataSourceHandler {
+    public DataSourceResponse handle(DataSourceRequest request) {
```
Comment (kkrik-es) [inline test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/datasource/DefaultPrimitiveTypesHandler.java:16]: Similarly, this looks like a good fit for inheritance?

File: test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/datasource/DataSourceResponse.java (hunk 3 above)
Comment (kkrik-es) [reply test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/datasource/DataSourceResponse.java:47]: Nm, `ObjectArrayGenerator` is defined separately.

**Review by kkrik-es — APPROVED** [review]
Nice, I much prefer this too. I was puzzled by the if-sequences, wonder if we can replace them with overriden methods. I can also give it a try once you submit it.

**Comment by lkts** [comment]
@kkrik-es i completely forgot that method overrides could work here, that should be much better

File: test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/datasource/DefaultPrimitiveTypesHandler.java (hunk 5 above)
Comment (lkts) [reply test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/datasource/DefaultPrimitiveTypesHandler.java:16]: Sorry i don't fully see what's the inheritance structure.

Comment (lkts) [reply test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/datasource/DefaultPrimitiveTypesHandler.java:16]: I think i figured it out.

File: test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/datasource/DefaultObjectGenerationHandler.java (hunk 4 above)
Comment (lkts) [reply test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/datasource/DefaultObjectGenerationHandler.java]: I think half the time is okay but i can change.

**Review by kkrik-es — APPROVED** [review]

**Comment by lkts** [comment]
@elasticmachine update branch

**Pull request merged**